src/core/lp_lexer.ml

```diff
@@ -0,0 +1,216 @@
+(** Lexer for Lambdapi syntax. Uses Sedlex, a Utf8 friendly lexer. Some helper
+    functions to check if a string conflicts with the syntax are also
+    provided. *)
+open Sedlexing
+open Pos
+
+type token =
+  | EOF
+  | L_PAREN | R_PAREN | L_SQ_BRACKET | R_SQ_BRACKET
+  | L_CU_BRACKET | R_CU_BRACKET
+  | APPLY | ARROW | AS | ASSERT | ASSERT_NOT | ASSIGN | ASSOC of Syntax.assoc
+  | AT
+  | BUILTIN | BACKQUOTE
+  | COLON | COMMA | COMPUTE | COMPUTE_TYPE | CONSTANT
+  | DEBUG_FLAGS of (bool * string)
+  (** Flags such as [+eiu] or [-eiu]. *)
+  | DEFINITION | DOLLAR | DOT
+  | EQUIV
+  | FAIL | FLAG | FLOAT of float | FOCUS
+  | ID of (string * bool)
+   (** Boolean is true if ident is escaped *)
+  | IN | INFERTYPE | INDUCTIVE | INFIX | INJECTIVE | INT of int | INTRO
+  | LAMBDA | LET
+  | OPEN
+  | PI | PREFIX | PRINT | PRIVATE | PROOF | PROOF_END of Syntax.p_proof_end
+  | PROOFTERM | PROTECTED | PROVER | PROVER_TIMEOUT
+  | QUESTION_MARK
+  | REFINE | REFL | REQUIRE | REWRITE | RULE
+  | SEMICOLON | SEQUENTIAL | SET | SIMPL | STRINGLIT of string
+  | SWITCH of bool
+  (** [on] or [off], for flags. *)
+  | SYMMETRY | SYMBOL
+  | THEOREM | TYPE | TURNSTILE
+  | UNIF_RULE
+  | VBAR
+  | VERBOSE
+  | WHY3 | WILD | WITH
+
+exception SyntaxError of strloc
+
+(** Lexer for Lambdapi syntax. *)
+module Lp_lexer : sig
+
+  val is_escaped : string -> bool
+  (** [is_escped s] is true if [s] is an escaped identifier, that is, an
+     identifier enclosed between [{|] and [|}]. *)
+
+  val is_identifier : string -> bool
+  (** [is_identifier s] is [true] if [s] is a valid identifier. *)
+
+  val is_debug_flag : string -> bool
+  (** [is_debug_flag s] is true if [s] is a debug flag. *)
+
+  val lexer : lexbuf -> unit -> token * Lexing.position * Lexing.position
+  (** [lexer buf] is a lexing function on buffer [buf] that can be passed to
+      a parser. *)
+end = struct
+  let digit = [%sedlex.regexp? '0' .. '9']
+  let integer = [%sedlex.regexp? Plus digit]
+
+  (* We define the set of UTF8 codepoints that make up identifiers. *)
+  let superscript =
+    [%sedlex.regexp? 0x2070 | 0x00b9 | 0x00b2 | 0x00b3 | 0x2074 .. 0x207c]
+  let subscript = [%sedlex.regexp? 0x208 .. 0x208c]
+  let supplemental_punctuation = [%sedlex.regexp? 0x2e00 .. 0x2e52]
+  let alphabet = [%sedlex.regexp? 'a' .. 'z' | 'A' .. 'Z']
+  let ascii_sub =
+    [%sedlex.regexp? '-' | '\'' | '&' | '^' | '\\' | '*'
+                   | '%' | '#' | '~']
+  (* NOTE: lambda character now counts as a letter, so a sequence λx is an
+     identifier, so [λx t u,] fails on unexpected token [,]. But [λ x t u,]
+     works. *)
+  let letter =
+    [%sedlex.regexp? lowercase | uppercase | ascii_sub
+                   | math | other_math | subscript | superscript
```

Comment:
What are math and other_math?

Reply:
Indeed what these categories are is not really clear. They ought to be defined in ISO standards, but it's a mess to find the precise definitions. The categories are defined on the github page of sedlex https://github.com/ocaml-community/sedlex

Reply:
Is it important to restrict the set of identifiers? Why not taking any character?

Reply:
The lexing became very slow, it must be linked to how sedlex works

Reply:
I see. Thanks to add a comment with the URL you mention so that we remember it.